### PR TITLE
adding support to render error responses per the graphql spec, closes #68

### LINKED ISSUE #68: Error handling for mutations
Hi,

is there any support for handling errors like specified in the latest graphQL spec?
https://graphql.github.io/graphql-spec/June2018/#sec-Errors

Currently we try to delete multiple entries in our database through one bulk mutation.
The problem is that we are uncertain how the graphql-response should look like, when 1 or more entries could not be deleted because, e.g. the current user has the wrong permissions.

Thanks in advance for your answer :)

### PR DESCRIPTION
**Pull request by drmaas**
adding support to render error responses per the graphql spec, closes #68

Need some guidance on:

* handling edge cases for default error response rendering
* adding tests to assert this works correctly for queries and mutations

### CODE REVIEW

**Review by jeggy — COMMENTED** [review]
The `ErrorFunction` looks great :ok_hand:  

Currently in the `SchemaConfiguration` we have a `wrapErrors` boolean. I think we should replace this configuration with a new one. This new configuration could be called `errorStrategy` and be of a enum class type with the options:
* `STANDARD` - Would be this new error handling you have built here
* `WRAP_EXCEPTION` - Which would be the same as `wrapErrors = true`
* `NOTHING` - Which would be the same as `wrapErrors = false`

File: kgraphql/src/main/kotlin/com/apurebase/kgraphql/schema/execution/DataLoaderPreparedRequestExecutor.kt [hunk 1]
```diff
@@ -361,15 +368,30 @@ class DataLoaderPreparedRequestExecutor(val schema: DefaultSchema) : RequestExec
...
-
             "data" toDeferredObj {
                 plan.forEach { node ->
-                    if (shouldInclude(ctx, node)) writeOperation(ctx, node, node.field as Field.Function<*, *>)
+                    if (shouldInclude(ctx, node))
+                        try {
+                            writeOperation(ctx, node, node.field as Field.Function<*, *>)
+                        } catch (e: GraphQLError) {
+                            exceptionMap[node] = e
+                        }
```
[the remark this answers is not in the document]
Comment (drmaas) [reply kgraphql/src/main/kotlin/com/apurebase/kgraphql/schema/execution/DataLoaderPreparedRequestExecutor.kt:385]: @jeggy I am not sure how to make this work inside of `fun FunctionWrapper<T>.invoke` - how would thrown exceptions be aggregated? Right now, it looks like the first error encountered will throw an exception and halt further processing. I am going to keep looking at it in the meantime.

**Comment by drmaas** [comment]
@jeggy I am going to close this - I don't have cycles to further investigate how error handling would work at holistic level, and we are thinking of moving to https://github.com/ExpediaGroup/graphql-kotlin/ internally. But, the branch will be kept on my fork if needed.

**Comment by jeggy** [comment]
> @jeggy I am going to close this - I don't have cycles to further investigate how error handling would work at holistic level, and we are thinking of moving to https://github.com/ExpediaGroup/graphql-kotlin/ internally. But, the branch will be kept on my fork if needed.

Sorry to hear that, and I do understand.
Thanks for your fork :+1: I will try to take over from here and to build proper built-in support for errors handling.
